transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java

```diff
@@ -287,38 +279,11 @@ protected void run() {
                         break;
 
                     case SelectStrategy.SELECT:
-                        strategy = epollWait(WAKEN_UP_UPDATER.getAndSet(this, 0) == 1);
-
-                        // 'wakenUp.compareAndSet(false, true)' is always evaluated
-                        // before calling 'selector.wakeup()' to reduce the wake-up
-                        // overhead. (Selector.wakeup() is an expensive operation.)
-                        //
-                        // However, there is a race condition in this approach.
-                        // The race condition is triggered when 'wakenUp' is set to
-                        // true too early.
-                        //
-                        // 'wakenUp' is set to true too early if:
-                        // 1) Selector is waken up between 'wakenUp.set(false)' and
-                        //    'selector.select(...)'. (BAD)
-                        // 2) Selector is waken up between 'selector.select(...)' and
-                        //    'if (wakenUp.get()) { ... }'. (OK)
-                        //
-                        // In the first case, 'wakenUp' is set to true and the
-                        // following 'selector.select(...)' will wake up immediately.
-                        // Until 'wakenUp' is set to false again in the next round,
-                        // 'wakenUp.compareAndSet(false, true)' will fail, and therefore
-                        // any attempt to wake up the Selector will fail, too, causing
-                        // the following 'selector.select(...)' call to block
-                        // unnecessarily.
-                        //
-                        // To fix this problem, we wake up the selector again if wakenUp
-                        // is true immediately after selector.select(...).
-                        // It is inefficient in that it wakes up the selector for both
-                        // the first case (BAD - wake-up required) and the second case
-                        // (OK - no wake-up required).
-
                         if (wakenUp == 1) {
-                            Native.eventFdWrite(eventFd.intValue(), 1L);
+                            wakenUp = 0;
+                        }
+                        if (!hasTasks()) {
+                            strategy = epollWait();
                         }
```

Comment:
After looking again at this PR I think it should be changed slightly to do this:

```
if (!hasTasks()) {
    strategy = epollWait();
} else {
    strategy = epollWaitNow():
}
```
The reason for this is that otherwise we will never attempt to do any IO if there is at least one task. This sounds not correct may result in timeouts etc due stale connections.

Reply:
Done, and added a comment.

Reply:
can we flip it around so there is no negation?

Reply:
I could go either way.  I generally prefer having the more likely branch first (so that the code more generally reads from top to bottom).

Reply:
Would be nice if Java had branch prediction macros :)

Reply:
Let's leave it as is with the most likely branch first then :+1:

Reply:
@normanmaurer are you sure this is needed? The same check would have already just been done right above, which will call `epollWaitNow()`: https://github.com/netty/netty/blob/799c9a9b7d5deb1b90525bf9ffc55179d0238795/transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java#L272

The additional `hasTasks()` is only needed here to avoid race with `wakenUp` change and task submission.

In fact we could in any case simplify slightly here to only check it if the prior check returned `false`:
```java
boolean hasTasks = hasTasks();
strategy = selectStrategy.calculateStrategy(selectNowSupplier, hasTasks); 
...
if (wakenUp == 1) {
    wakenUp = 0;
}
if (!hasTasks && !hasTasks()) {
    strategy = epollWait();
}
```

Reply:
@njhill you are right 🤦‍♂ ... I did miss the usage of `selectNowSupplier`. @carl-mastrangelo please revert my suggestion as it is not needed as correctly pointed out by @njhill

Reply:
Flipped